File: scripts/raw_news_contract.py
from __future__ import annotations

import hashlib
import json
import re
import subprocess
from dataclasses import dataclass
from functools import lru_cache
from html import unescape
from pathlib import Path
from typing import Any
from urllib.parse import urlparse

CONFIG_PATH = Path(__file__).resolve().parent.parent / "config" / "news_raw_contract.json"
EXTRACTOR_SCRIPT_PATH = Path(__file__).resolve().parent / "extract-article.mjs"

CLICKBAIT_PATTERNS: tuple[re.Pattern[str], ...] = (
    re.compile(r"you won't believe", re.IGNORECASE),
    re.compile(r"\d+\s+stocks?\s+to\s+buy", re.IGNORECASE),
    re.compile(r"next\s+(big|huge)\s+thing", re.IGNORECASE),
    re.compile(r"get rich", re.IGNORECASE),
    re.compile(r"millionaire", re.IGNORECASE),
    re.compile(r"(bitcoin|ethereum|crypto|altcoin|memecoin|dogecoin)", re.IGNORECASE),
    re.compile(r"(nft|web3|metaverse)", re.IGNORECASE),
    re.compile(r"penny stock", re.IGNORECASE),
    re.compile(r"passive income", re.IGNORECASE),
    re.compile(r"(horoscope|zodiac|astrology)", re.IGNORECASE),
    re.compile(r"celebrity", re.IGNORECASE),
)

HTML_TAG_RE = re.compile(r"<[^>]+>")
WHITESPACE_RE = re.compile(r"\s+")
NON_ALNUM_RE = re.compile(r"[^a-z0-9]+")
NUMBER_SIGNAL_RE = re.compile(r"\b\d+(?:[.,]\d+)?\s*(?:%|bp|bps|k|m|b|points?)\b", re.IGNORECASE)


@dataclass(frozen=True)
class TopicSpec:
    code: str
    label: str
    family: str
    econ_category: str | None
    tags: tuple[str, ...]
    google_query: str
    keywords: tuple[str, ...]


@dataclass(frozen=True)
class RawNewsContract:
    scoring_version: str
    watchlist_symbols: tuple[str, ...]
    blocked_domains: frozenset[str]
    trusted_domains: frozenset[str]
    premier_domains: frozenset[str]
    newsfilter_source_allowlist: frozenset[str]
    finnhub_categories: tuple[str, ...]
    benchmark_weights: dict[str, float]
    benchmark_terms: dict[str, tuple[str, ...]]
    topics: tuple[TopicSpec, ...]

    @property
    def topic_codes(self) -> tuple[str, ...]:
        return tuple(topic.code for topic in self.topics)

    @property
    def topic_map(self) -> dict[str, TopicSpec]:
        return {topic.code: topic for topic in self.topics}


@lru_cache(maxsize=1)
def load_raw_news_contract() -> RawNewsContract:
    payload = json.loads(CONFIG_PATH.read_text())
    topics = tuple(
        TopicSpec(
            code=entry["code"],
            label=entry["label"],
            family=entry["family"],
            econ_category=entry.get("econ_category"),
            tags=tuple(entry.get("tags", [])),
            google_query=entry["google_query"],
            keywords=tuple(entry.get("keywords", [])),
        )
        for entry in payload["topics"]
    )
    benchmark_terms = {
        key: tuple(values)
        for key, values in payload["benchmark_terms"].items()
    }
    return RawNewsContract(
        scoring_version=payload["scoring_version"],
        watchlist_symbols=tuple(payload["watchlist_symbols"]),
        blocked_domains=frozenset(payload["blocked_domains"]),
        trusted_domains=frozenset(payload["trusted_domains"]),
        premier_domains=frozenset(payload["premier_domains"]),
        newsfilter_source_allowlist=frozenset(payload["newsfilter_source_allowlist"]),
        finnhub_categories=tuple(payload["finnhub_categories"]),
        benchmark_weights={key: float(value) for key, value in payload["benchmark_weights"].items()},
        benchmark_terms=benchmark_terms,
        topics=topics,
    )


def normalize_text(value: str | None) -> str:
    if not value:
        return ""
    return WHITESPACE_RE.sub(" ", unescape(value)).strip()


def strip_html_tags(value: str | None) -> str:
    return normalize_text(HTML_TAG_RE.sub(" ", value or ""))


def extract_domain(url: str | None) -> str | None:
    if not url:
        return None
    try:
        hostname = urlparse(url).hostname or ""
    except ValueError:
        return None
    hostname = hostname.lower()
    if hostname.startswith("www."):
        hostname = hostname[4:]
    return hostname or None


def normalize_title_for_dedupe(title: str) -> str:
    lowered = normalize_text(title).lower()
    lowered = NON_ALNUM_RE.sub(" ", lowered)
    return WHITESPACE_RE.sub(" ", lowered).strip()


def published_minute_iso(iso_value: str) -> str:
    trimmed = iso_value[:16]
    if trimmed.endswith(":"):
        trimmed = trimmed[:-1]
    return f"{trimmed}:00+00:00"


def build_article_key(*parts: str) -> str:
    base = "|".join(part for part in parts if part)
    return hashlib.sha256(base.encode("utf-8")).hexdigest()


def build_dedupe_key(normalized_title: str, publisher_domain: str | None, published_minute: str) -> str:
    return build_article_key(normalized_title, publisher_domain or "", published_minute)


def is_junk(title: str, summary: str, publisher_domain: str | None, contract: RawNewsContract) -> bool:
    if publisher_domain in contract.blocked_domains:
        return True
    if publisher_domain in contract.trusted_domains:
        return False
    haystack = f"{title} {summary}"
    for pattern in CLICKBAIT_PATTERNS:
        if pattern.search(haystack):
            return True
    return len(normalize_text(title)) < 15


def matched_keywords(title: str, summary: str, keywords: tuple[str, ...]) -> tuple[str, ...]:
    haystack = f"{title} {summary}".lower()
    return tuple(keyword for keyword in keywords if keyword.lower() in haystack)


def extract_watchlist_symbols(
    contract: RawNewsContract,
    text: str,
    explicit_symbols: list[str] | tuple[str, ...] | None = None,
) -> tuple[str, ...]:
    haystack = f" {normalize_text(text).upper()} "
    matched: list[str] = []
    explicit = {symbol.upper() for symbol in (explicit_symbols or [])}
    for symbol in contract.watchlist_symbols:
        if symbol.upper() in explicit:
            matched.append(symbol)
            continue
        if f" {symbol.upper()} " in haystack:
            matched.append(symbol)
    return tuple(dict.fromkeys(matched))


def count_term_matches(text: str, terms: tuple[str, ...]) -> int:
    lowered = normalize_text(text).lower()
    return sum(1 for term in terms if term.lower() in lowered)


def count_number_signals(text: str) -> int:
    return len(NUMBER_SIGNAL_RE.findall(text))


def _bounded_score(raw_value: float, ceiling: float) -> float:
    return max(0.0, min(raw_value / ceiling, 1.0))


def score_article(
    *,
    provider: str,
    publisher_domain: str | None,
    topic: TopicSpec,
    title: str,
    summary: str | None,
    body_text: str | None,
    image_url: str | None,
    explicit_symbols: list[str] | tuple[str, ...] | None = None,
    matched_topic_keywords: tuple[str, ...] = (),
) -> dict[str, Any]:
    contract = load_raw_news_contract()
    text = normalize_text(f"{title} {summary or ''} {body_text or ''}")
    identified_symbols = extract_watchlist_symbols(contract, text, explicit_symbols)
    market_match_count = count_term_matches(text, contract.benchmark_terms["market_context"])
    macro_match_count = count_term_matches(text, contract.benchmark_terms["macro_specificity"])
    technical_match_count = count_term_matches(text, contract.benchmark_terms["technical_specificity"])
    cross_asset_match_count = count_term_matches(text, contract.benchmark_terms["cross_asset_context"])
    number_match_count = count_number_signals(text)

    if publisher_domain in contract.premier_domains:
        source_quality_score = 1.0
    elif publisher_domain in contract.trusted_domains:
        source_quality_score = 0.82
    elif publisher_domain:
        source_quality_score = 0.55
    else:
        source_quality_score = 0.45

    market_relevance_score = min(
        1.0,
        0.45 + _bounded_score(len(matched_topic_keywords), 4.0) * 0.35 + _bounded_score(market_match_count, 4.0) * 0.20,
    )
    macro_specificity_score = min(
        1.0,
        _bounded_score(macro_match_count, 4.0) * 0.65 + _bounded_score(number_match_count, 3.0) * 0.35,
    )
    technical_specificity_score = min(
        1.0,
        _bounded_score(technical_match_count, 4.0) * 0.70 + _bounded_score(number_match_count, 4.0) * 0.30,
    )
    cross_asset_context_score = _bounded_score(cross_asset_match_count, 4.0)
    image_presence_score = 1.0 if image_url else 0.0
    watchlist_relevance_score = _bounded_score(len(identified_symbols), 2.0)

    weights = contract.benchmark_weights
    benchmark_fit_score = (
        source_quality_score * weights["source_quality"]
        + market_relevance_score * weights["market_relevance"]
        + macro_specificity_score * weights["macro_specificity"]
        + technical_specificity_score * weights["technical_specificity"]
        + cross_asset_context_score * weights["cross_asset_context"]
        + image_presence_score * weights["image_presence"]
        + watchlist_relevance_score * weights["watchlist_relevance"]
    )

    reason_flags: list[str] = [f"topic:{topic.code}", f"family:{topic.family}", f"provider:{provider}"]
    if publisher_domain in contract.premier_domains:
        reason_flags.append("premier_source")
    elif publisher_domain in contract.trusted_domains:
        reason_flags.append("trusted_source")
    if matched_topic_keywords:
        reason_flags.append("topic_keyword_match")
    if market_match_count > 0:
        reason_flags.append("market_context")
    if macro_match_count > 0:
        reason_flags.append("macro_context")
    if number_match_count > 0:
        reason_flags.append("exact_numbers")
    if technical_match_count > 0:
        reason_flags.append("technical_levels")
    if cross_asset_match_count > 0:
        reason_flags.append("cross_asset_context")
    if image_url:
        reason_flags.append("image_present")
    if identified_symbols:
        reason_flags.append("watchlist_symbol_match")

    reasoning_confidence = min(
        1.0,
        0.25
        + _bounded_score(len(reason_flags), 8.0) * 0.45
        + benchmark_fit_score * 0.30,
    )

    return {
        "provider": provider,
        "identified_symbols": list(identified_symbols),
        "reason_flags": list(dict.fromkeys(reason_flags)),
        "source_quality_score": round(source_quality_score, 4),
        "market_relevance_score": round(market_relevance_score, 4),
        "macro_specificity_score": round(macro_specificity_score, 4),
        "technical_specificity_score": round(technical_specificity_score, 4),
        "cross_asset_context_score": round(cross_asset_context_score, 4),
        "image_presence_score": round(image_presence_score, 4),
        "watchlist_relevance_score": round(watchlist_relevance_score, 4),
        "reasoning_confidence": round(reasoning_confidence, 4),
        "benchmark_fit_score": round(benchmark_fit_score, 4),
        "evidence": {
            "matched_topic_keywords": list(matched_topic_keywords),
            "market_term_hits": market_match_count,
            "macro_term_hits": macro_match_count,
            "technical_term_hits": technical_match_count,
            "cross_asset_term_hits": cross_asset_match_count,
            "number_signal_hits": number_match_count,
            "image_present": bool(image_url),
        },
    }


def extract_article_with_node(url: str) -> dict[str, Any]:
    result = subprocess.run(
        ["node", str(EXTRACTOR_SCRIPT_PATH), url],
        capture_output=True,
        check=False,
        text=True,
    )
    output = (result.stdout or "").strip()
    if not output:
        raise RuntimeError(result.stderr.strip() or "article extractor returned no output")
    return json.loads(output)

File: supabase/functions/_shared/raw-news-contract.ts
// Raw news contract, scoring, and utilities.
// Ported from lib/news/raw-news-contract.ts:
//   - import { createHash } from "crypto" → import { createHash } from "node:crypto"
//   - import contractJson from "@/config/..." → relative JSON with assert { type: "json" }

import { createHash } from "node:crypto";
import contractJson from "./news-raw-contract.json" assert { type: "json" };

export type TopicSpec = {
  code: string;
  label: string;
  family: string;
  econ_category: string | null;
  tags: string[];
  google_query: string;
  keywords: string[];
};

type BenchmarkWeights = {
  source_quality: number;
  market_relevance: number;
  macro_specificity: number;
  technical_specificity: number;
  cross_asset_context: number;
  image_presence: number;
  watchlist_relevance: number;
};

export type ArticleScore = {
  identifiedSymbols: string[];
  reasonFlags: string[];
  sourceQualityScore: number;
  marketRelevanceScore: number;
  macroSpecificityScore: number;
  technicalSpecificityScore: number;
  crossAssetContextScore: number;
  imagePresenceScore: number;
  watchlistRelevanceScore: number;
  reasoningConfidence: number;
  benchmarkFitScore: number;
  evidence: {
    matchedTopicKeywords: string[];
    marketTermHits: number;
    macroTermHits: number;
    technicalTermHits: number;
    crossAssetTermHits: number;
    numberSignalHits: number;
    imagePresent: boolean;
  };
};

export const RAW_NEWS_CONTRACT = {
  scoringVersion: contractJson.scoring_version,
  watchlistSymbols: contractJson.watchlist_symbols,
  blockedDomains: new Set(contractJson.blocked_domains),
  trustedDomains: new Set(contractJson.trusted_domains),
  premierDomains: new Set(contractJson.premier_domains),
  newsfilterSourceAllowlist: new Set(contractJson.newsfilter_source_allowlist),
  finnhubCategories: contractJson.finnhub_categories,
  finnhubCompanyNewsSymbols: contractJson.finnhub_company_news_symbols,
  minBenchmarkFitScore: contractJson.min_benchmark_fit_score,
  benchmarkWeights: contractJson.benchmark_weights as BenchmarkWeights,
  benchmarkTerms: contractJson.benchmark_terms,
  topics: contractJson.topics as TopicSpec[],
} as const;

export const RAW_NEWS_TOPIC_MAP = new Map(
  RAW_NEWS_CONTRACT.topics.map((topic) => [topic.code, topic]),
);

const CLICKBAIT_PATTERNS = [
  /you won't believe/i,
  /\d+\s+stocks?\s+to\s+buy/i,
  /next\s+(big|huge)\s+thing/i,
  /get rich/i,
  /millionaire/i,
  /(bitcoin|ethereum|crypto|altcoin|memecoin|dogecoin)/i,
  /(nft|web3|metaverse)/i,
  /penny stock/i,
  /passive income/i,
  /(horoscope|zodiac|astrology)/i,
  /celebrity/i,
];

const NUMBER_SIGNAL_RE = /\b\d+(?:[.,]\d+)?\s*(?:%|bp|bps|k|m|b|points?)\b/gi;

export function normalizeText(value: string | null | undefined): string {
  return (value ?? "").replace(/\s+/g, " ").trim();
}

export function stripHtmlTags(value: string | null | undefined): string {
  return normalizeText((value ?? "").replace(/<[^>]+>/g, " "));
}

export function extractDomain(url: string | null | undefined): string | null {
  if (!url) {
    return null;
  }
  try {
    const hostname = new URL(url).hostname.toLowerCase().replace(/^www\./, "");
    return hostname || null;
  } catch {
    return null;
  }
}

export function normalizeTitleForDedupe(title: string): string {
  return normalizeText(title).toLowerCase().replace(/[^a-z0-9]+/g, " ");
}

export function buildArticleKey(...parts: Array<string | null | undefined>): string {
  const base = parts.filter(Boolean).join("|");
  return createHash("sha256").update(base).digest("hex");
}

export function publishedMinuteIso(value: string): string {
  const minute = new Date(value);
  minute.setUTCSeconds(0, 0);
  return minute.toISOString();
}

export function buildDedupeKey(
  normalizedTitle: string,
  publisherDomain: string | null,
  publishedMinute: string,
): string {
  return buildArticleKey(normalizedTitle, publisherDomain ?? "", publishedMinute);
}

export function matchedKeywords(title: string, bodyText: string, keywords: string[]): string[] {
  const haystack = `${title} ${bodyText}`.toLowerCase();
  return keywords.filter((keyword) => haystack.includes(keyword.toLowerCase()));
}

export function extractWatchlistSymbols(text: string, explicitSymbols: string[] = []): string[] {
  const haystack = ` ${normalizeText(text).toUpperCase()} `;
  const explicit = new Set(explicitSymbols.map((symbol) => symbol.toUpperCase()));
  const matches: string[] = [];
  for (const symbol of RAW_NEWS_CONTRACT.watchlistSymbols) {
    if (explicit.has(symbol.toUpperCase()) || haystack.includes(` ${symbol.toUpperCase()} `)) {
      matches.push(symbol);
    }
  }
  return [...new Set(matches)];
}

export function isJunk(
  title: string,
  summary: string,
  publisherDomain: string | null,
): boolean {
  if (publisherDomain && RAW_NEWS_CONTRACT.blockedDomains.has(publisherDomain)) {
    return true;
  }
  if (publisherDomain && RAW_NEWS_CONTRACT.trustedDomains.has(publisherDomain)) {
    return false;
  }
  const haystack = `${title} ${summary}`;
  return CLICKBAIT_PATTERNS.some((pattern) => pattern.test(haystack)) || normalizeText(title).length < 15;
}

function countTermMatches(text: string, terms: string[]): number {
  const lowered = normalizeText(text).toLowerCase();
  return terms.reduce((count, term) => count + (lowered.includes(term.toLowerCase()) ? 1 : 0), 0);
}

function countNumberSignals(text: string): number {
  return (text.match(NUMBER_SIGNAL_RE) ?? []).length;
}

function boundedScore(rawValue: number, ceiling: number): number {
  return Math.max(0, Math.min(rawValue / ceiling, 1));
}

export function scoreArticle(params: {
  provider: string;
  publisherDomain: string | null;
  topic: TopicSpec;
  title: string;
  summary: string | null;
  bodyText: string | null;
  imageUrl: string | null;
  explicitSymbols?: string[];
  matchedTopicKeywords?: string[];
}): ArticleScore {
  const {
    provider,
    publisherDomain,
    topic,
    title,
    summary,
    bodyText,
    imageUrl,
    explicitSymbols = [],
    matchedTopicKeywords = [],
  } = params;

  const fullText = normalizeText(`${title} ${summary ?? ""} ${bodyText ?? ""}`);
  const identifiedSymbols = extractWatchlistSymbols(fullText, explicitSymbols);
  const marketTermHits = countTermMatches(fullText, [...RAW_NEWS_CONTRACT.benchmarkTerms.market_context]);
  const macroTermHits = countTermMatches(fullText, [...RAW_NEWS_CONTRACT.benchmarkTerms.macro_specificity]);
  const technicalTermHits = countTermMatches(fullText, [...RAW_NEWS_CONTRACT.benchmarkTerms.technical_specificity]);
  const crossAssetTermHits = countTermMatches(fullText, [...RAW_NEWS_CONTRACT.benchmarkTerms.cross_asset_context]);
  const numberSignalHits = countNumberSignals(fullText);

  let sourceQualityScore = 0.45;
  if (publisherDomain && RAW_NEWS_CONTRACT.premierDomains.has(publisherDomain)) {
    sourceQualityScore = 1;
  } else if (publisherDomain && RAW_NEWS_CONTRACT.trustedDomains.has(publisherDomain)) {
    sourceQualityScore = 0.82;
  } else if (publisherDomain) {
    sourceQualityScore = 0.55;
  }

  const marketRelevanceScore = Math.min(
    1,
    0.45 + boundedScore(matchedTopicKeywords.length, 4) * 0.35 + boundedScore(marketTermHits, 4) * 0.2,
  );
  const macroSpecificityScore = Math.min(
    1,
    boundedScore(macroTermHits, 4) * 0.65 + boundedScore(numberSignalHits, 3) * 0.35,
  );
  const technicalSpecificityScore = Math.min(
    1,
    boundedScore(technicalTermHits, 4) * 0.7 + boundedScore(numberSignalHits, 4) * 0.3,
  );
  const crossAssetContextScore = boundedScore(crossAssetTermHits, 4);
  const imagePresenceScore = imageUrl ? 1 : 0;
  const watchlistRelevanceScore = boundedScore(identifiedSymbols.length, 2);

  const weights = RAW_NEWS_CONTRACT.benchmarkWeights;
  const benchmarkFitScore =
    sourceQualityScore * weights.source_quality +
    marketRelevanceScore * weights.market_relevance +
    macroSpecificityScore * weights.macro_specificity +
    technicalSpecificityScore * weights.technical_specificity +
    crossAssetContextScore * weights.cross_asset_context +
    imagePresenceScore * weights.image_presence +
    watchlistRelevanceScore * weights.watchlist_relevance;

  const reasonFlags = [`topic:${topic.code}`, `family:${topic.family}`, `provider:${provider}`];
  if (publisherDomain && RAW_NEWS_CONTRACT.premierDomains.has(publisherDomain)) {
    reasonFlags.push("premier_source");
  } else if (publisherDomain && RAW_NEWS_CONTRACT.trustedDomains.has(publisherDomain)) {
    reasonFlags.push("trusted_source");
  }
  if (matchedTopicKeywords.length > 0) {
    reasonFlags.push("topic_keyword_match");
  }
  if (marketTermHits > 0) {
    reasonFlags.push("market_context");
  }
  if (macroTermHits > 0) {
    reasonFlags.push("macro_context");
  }
  if (numberSignalHits > 0) {
    reasonFlags.push("exact_numbers");
  }
  if (technicalTermHits > 0) {
    reasonFlags.push("technical_levels");
  }
  if (crossAssetTermHits > 0) {
    reasonFlags.push("cross_asset_context");
  }
  if (imageUrl) {
    reasonFlags.push("image_present");
  }
  if (identifiedSymbols.length > 0) {
    reasonFlags.push("watchlist_symbol_match");
  }

  const reasoningConfidence = Math.min(
    1,
    0.25 + boundedScore(reasonFlags.length, 8) * 0.45 + benchmarkFitScore * 0.3,
  );

  return {
    identifiedSymbols,
    reasonFlags: [...new Set(reasonFlags)],
    sourceQualityScore: Number(sourceQualityScore.toFixed(4)),
    marketRelevanceScore: Number(marketRelevanceScore.toFixed(4)),
    macroSpecificityScore: Number(macroSpecificityScore.toFixed(4)),
    technicalSpecificityScore: Number(technicalSpecificityScore.toFixed(4)),
    crossAssetContextScore: Number(crossAssetContextScore.toFixed(4)),
    imagePresenceScore: Number(imagePresenceScore.toFixed(4)),
    watchlistRelevanceScore: Number(watchlistRelevanceScore.toFixed(4)),
    reasoningConfidence: Number(reasoningConfidence.toFixed(4)),
    benchmarkFitScore: Number(benchmarkFitScore.toFixed(4)),
    evidence: {
      matchedTopicKeywords,
      marketTermHits,
      macroTermHits,
      technicalTermHits,
      crossAssetTermHits,
      numberSignalHits,
      imagePresent: Boolean(imageUrl),
    },
  };
}
